ЗАДАНИЕ 6. ОЦЕНКА КАЧЕСТВА МОДЕЛЕЙ

Раздел 1. Разбиение выборки на обучающий и тестовый набор

2.	Загрузите данные файла pima-indians-diabetes.data.csv с помощью функции read_csv() из модуля pandas. Названия столбцов укажите следующими: names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']. Посмотрите на первые 5 наблюдений с помощью функции head().

In [1]:
import pandas as pd
import numpy as np
url='https://raw.githubusercontent.com/kintis/HomeWork/main/pima-indians-diabetes.data.csv'
names=['preg','plas','pres','skin','test','mass','pedi','age','class']
df=pd.read_csv(url, names=names)
df.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   test    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


3.	Сохраните в переменную array значения признаков, обратившись к атрибуту values.

In [3]:
array=df.values
array

array([[  6.   , 148.   ,  72.   , ...,   0.627,  50.   ,   1.   ],
       [  1.   ,  85.   ,  66.   , ...,   0.351,  31.   ,   0.   ],
       [  8.   , 183.   ,  64.   , ...,   0.672,  32.   ,   1.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,   0.245,  30.   ,   0.   ],
       [  1.   , 126.   ,  60.   , ...,   0.349,  47.   ,   1.   ],
       [  1.   ,  93.   ,  70.   , ...,   0.315,  23.   ,   0.   ]])

4.	Выделите в переменной array часть, которая относится к факторным признакам, назовите ее X, и часть, которая относится к результативному признаку, - Y. Прокомментируйте содержимое переменных X и Y.

In [4]:
X, y = array[:, :-1], array[:, -1:]
X, y 

(array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
        [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
        [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
        ...,
        [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
        [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
        [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]]),
 array([[1.],
        [0.],
        [1.],
        [0.],
        [1.],
        [0.],
        [1.],
        [0.],
        [1.],
        [1.],
        [0.],
        [1.],
        [0.],
        [1.],
        [1.],
        [1.],
        [1.],
        [1.],
        [0.],
        [1.],
        [0.],
        [0.],
        [1.],
        [1.],
        [1.],
        [1.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [1.],
        [1.],
        [0.

5.	Разбейте выборку на обучающую и отложенную с помощью функции train_test_split(X, Y, test_size=test_size, random_state= random_state) из модуля sklearn.model_selection. Результат запишите в переменные X_train, X_test, Y_train, Y_test. Запишите в переменную test_size долю отложенной (тестовой) выборки. Пусть она составляет 0,33. Задайте в переменной random_state ядро случайных чисел для воспроизводимости результатов. Пусть random_state = 7.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=7, shuffle=True)
X_train, X_test, y_train, y_test



(array([[  3.   , 102.   ,  44.   , ...,  30.8  ,   0.4  ,  26.   ],
        [  1.   ,  77.   ,  56.   , ...,  33.3  ,   1.251,  24.   ],
        [  9.   , 124.   ,  70.   , ...,  35.4  ,   0.282,  34.   ],
        ...,
        [  0.   ,  57.   ,  60.   , ...,  21.7  ,   0.735,  67.   ],
        [  1.   , 105.   ,  58.   , ...,  24.3  ,   0.187,  21.   ],
        [  8.   , 179.   ,  72.   , ...,  32.7  ,   0.719,  36.   ]]),
 array([[  1.   ,  90.   ,  62.   , ...,  27.2  ,   0.58 ,  24.   ],
        [  7.   , 181.   ,  84.   , ...,  35.9  ,   0.586,  51.   ],
        [ 13.   , 152.   ,  90.   , ...,  26.8  ,   0.731,  43.   ],
        ...,
        [  4.   , 118.   ,  70.   , ...,  44.5  ,   0.904,  26.   ],
        [  7.   , 152.   ,  88.   , ...,  50.   ,   0.337,  36.   ],
        [  7.   , 168.   ,  88.   , ...,  38.2  ,   0.787,  40.   ]]),
 array([[0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
  

In [35]:
print(len(X_train), len(X_test), len(y_train), len(y_test))

514 254 514 254


6.	Создайте модель логистической регрессии с помощью функции LogisticRegression(solver='liblinear') из модуля sklearn.linear_model.
7.	Обучите ее на обучающей выборке с помощью функции fit().



In [6]:
from sklearn.linear_model import LogisticRegression
estimator=LogisticRegression(solver='liblinear')
estimator.fit(X_train, y_train)



c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(solver='liblinear')

8.	Оцените качество построенной модели, рассчитав на тестовом наборе долю правильных ответов (accuracy) с помощью функции score(). Запишите результат в переменную result и выведите его на экран.

In [7]:
result=estimator.score(X_test, y_test)
result

0.7559055118110236

9.	Создайте объект разбиения на фолды с помощью функции KFold(n_splits= n_splits, random_state = random_state, shuffle=True) из модуля sklearn.model_selection. Результат запишите в переменную kfold. На вход данной функции подайте следующую информацию: 1) количество частей, на которые должна разбиваться выборка (n_splits), - пусть 10; 2) счетчик случайных чисел (random_state) – пусть 7; 3) необходимость перемешивания данных выборки при каждой оценке, задав параметр shuffle=True.  

In [8]:
from sklearn.model_selection import KFold
kf=KFold(n_splits=10, shuffle=True, random_state=7)
kf.get_n_splits(X)



10

10.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold) из модуля sklearn.model_selection. Запишите результат в переменную results.

In [9]:
from sklearn.model_selection import cross_val_score
results=cross_val_score(estimator, X, y,cv=kf)
results


c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

array([0.83116883, 0.71428571, 0.71428571, 0.77922078, 0.79220779,
       0.76623377, 0.68831169, 0.85714286, 0.80263158, 0.76315789])

11.	Выведите на экран среднюю долю правильных ответов (accuracy) и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().

In [10]:
results.mean()

0.7708646616541353

In [11]:
results.std()

0.05090500786917546

12.	Создайте объект разбиения на фолды в пропорции «1:прочее» с помощью функции LeaveOneOut() из модуля sklearn.model_selection. Результат запишите в переменную loocv.

In [12]:
from sklearn.model_selection import LeaveOneOut
loocv=LeaveOneOut()
loocv.get_n_splits(X)


768

13.	Проведите кросс-валидацию с одним наблюдением для тестирования результата (Leave One Out Cross-Validation), воспользовавшись функцией cross_val_score(model, X, Y, cv=loocv) из модуля sklearn.model_selection. Запишите результат в переменную results.
14.	Выведите на экран среднюю долю правильных ответов (accuracy) и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().

In [17]:
results=cross_val_score(estimator, X, y,cv=loocv)
print(results.mean())
print(results.std())

c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

0.7682291666666666
0.42196340380334596


c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

15.	Создайте объект многократного случайного разбиения на фолды с помощью функции ShuffleSplit(n_splits=n_splits, test_size=test_size, random_state=seed) из модуля sklearn.model_selection. Результат запишите в переменную kfold. На вход данной функции подайте следующую информацию: 1) количество частей, на которые должна разбиваться выборка (n_splits), - пусть 10; 2) долю тестового набора (test_size) – пусть 0,33; 3) счетчик случайных чисел (random_state) – пусть 7.

In [19]:
from sklearn.model_selection import ShuffleSplit
rs = ShuffleSplit(n_splits=10, test_size=.33, random_state=7)
rs.get_n_splits(X)



10

16.	Проведите многократное случайное разбиение выборки на обучающую и тестовую, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold) из модуля sklearn.model_selection. Запишите результат в переменную results.

In [21]:
results =cross_val_score(estimator, X, y,cv=rs)
results

c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

array([0.75590551, 0.7519685 , 0.78346457, 0.76771654, 0.76377953,
       0.78346457, 0.7519685 , 0.79527559, 0.75984252, 0.73622047])

17.	Выведите на экран среднюю долю правильных ответов (accuracy) и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().

In [22]:
print(results.mean(), results.std())

0.7649606299212598 0.016983980007970873


Раздел 2. Метрики качества моделей классификации

18.	Создайте объект разбиения на фолды с помощью функции KFold(n_splits= n_splits, random_state = random_state, shuffle=True) из модуля sklearn.model_selection. Результат запишите в переменную kfold. На вход данной функции подайте следующую информацию: 1) количество частей, на которые должна разбиваться выборка (n_splits), - пусть 10; 2) счетчик случайных чисел (random_state) – пусть 7; 3) необходимость перемешивания данных выборки при каждой оценке, задав параметр shuffle=True.

In [23]:
from sklearn.model_selection import KFold
kf=KFold(n_splits=10, shuffle=True, random_state=7)
kf.get_n_splits(X)

10

19.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'accuracy'. Запишите результат в переменную results.
20.	Выведите на экран среднюю долю правильных ответов (accuracy) и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().


In [26]:
from sklearn.model_selection import cross_val_score
results=cross_val_score(estimator, X, y,cv=kf, scoring='accuracy')
print(results)
print(results.mean(), results.std())

[0.83116883 0.71428571 0.71428571 0.77922078 0.79220779 0.76623377
 0.68831169 0.85714286 0.80263158 0.76315789]
0.7708646616541353 0.05090500786917546


c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

21.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'neg_log_loss'. Разбиение на фолды (cv) задайте с помощью функции KFold() из модуля sklearn.model_selection, как в п. 18. Запишите результат в переменную results.
22.	Выведите на экран среднее значение логистической функции потерь, взятое с отрицательным знаком, и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().

In [27]:
kf=KFold(n_splits=10, shuffle=True, random_state=7)
results=cross_val_score(estimator, X, y,cv=kf, scoring='neg_log_loss')
print(results)
print(results.mean(), results.std())

[-0.45790068 -0.54106124 -0.5276601  -0.44801672 -0.48591327 -0.48252048
 -0.54906198 -0.44690861 -0.44645478 -0.55069593]
-0.4936193783449746 0.04205414143239438


c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

23.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'roc_auc'. Разбиение на фолды (cv) задайте с помощью функции KFold() из модуля sklearn.model_selection, как в п. 18. Запишите результат в переменную results.
24.	Выведите на экран среднюю площадь под ROC кривой и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std().


In [29]:
kf=KFold(n_splits=10, shuffle=True, random_state=7)
results=cross_val_score(estimator, X, y,cv=kf, scoring='roc_auc')
print(results)
print(results.mean(), results.std())


[0.90532959 0.76998492 0.77851852 0.8510467  0.81555556 0.81384615
 0.77514368 0.9214876  0.828125   0.79784946]
0.8256887180936934 0.050069703191417424


c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

25.	 Найдите расчетные значения по модели логистической регрессии на тестовой выборке с помощью функции predict(). Результат запишите в переменную predicted.

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=7, shuffle=True)
from sklearn.linear_model import LogisticRegression
estimator=LogisticRegression(solver='liblinear')
estimator.fit(X_train, y_train)
predicted = estimator.predict(X_test)




c:\Users\Iuliia\miniconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


26.	Посмотрите на матрицу ошибок прогнозирования на тестовой выборке для модели логистической регрессии, построенной и обученной в пп. 7-9. Воспользуйтесь при этом функцией confusion_matrix() из модуля sklearn.metrics. На вход этой функции подайте фактические значения результативного признака на тестовом наборе (Y_test) и расчетные значения по модели (predicted). Выведите на экран полученный результат.

In [34]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predicted)


array([[141,  21],
       [ 41,  51]], dtype=int64)

27.	Прокомментируйте полученный результат.

***141 - TP, 51 - TN, 21 - FP, 41 - FN***

28.	Создайте матрицу ошибок с помощью функции classification_report() из модуля sklearn.metrics. Результат запишите в переменную report. На вход этой функции подайте фактические значения результативного признака на тестовом наборе (Y_test) и расчетные значения по модели (predicted). Выведите на экран полученный результат, воспользовавшись функцией print().

In [37]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predicted))


              precision    recall  f1-score   support

         0.0       0.77      0.87      0.82       162
         1.0       0.71      0.55      0.62        92

    accuracy                           0.76       254
   macro avg       0.74      0.71      0.72       254
weighted avg       0.75      0.76      0.75       254



29.	Прокомментируйте полученный результат.

***выборка несбалансированная (162/92)***


Раздел 3. Метрики качества моделей регрессии

30.	Загрузите данные файла housing.csv с помощью функции read_csv(filename, delim_whitespace=True, names=names) из модуля pandas. Названия столбцов укажите следующими: names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT', 'MEDV']. Посмотрите на первые 5 наблюдений с помощью функции head().

In [49]:
url='https://raw.githubusercontent.com/kintis/HomeWork/main/housing.csv'
names=['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT', 'MEDV']
df_=pd.read_csv(url, delim_whitespace=True, names=names)
df_.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [50]:
df_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [51]:
df_.describe()


CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

31.	Сохраните в переменную array значения признаков, обратившись к атрибуту values.
32.	Выделите в переменной array часть, которая относится к факторным признакам, назовите ее X, и часть, которая относится к результативному признаку, - Y. Прокомментируйте содержимое переменных X и Y.


In [58]:
array=df_.values
X=array[:, :-1]
y=array[:, -1:]
print(X)
print(y)

[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]
[[24. ]
 [21.6]
 [34.7]
 [33.4]
 [36.2]
 [28.7]
 [22.9]
 [27.1]
 [16.5]
 [18.9]
 [15. ]
 [18.9]
 [21.7]
 [20.4]
 [18.2]
 [19.9]
 [23.1]
 [17.5]
 [20.2]
 [18.2]
 [13.6]
 [19.6]
 [15.2]
 [14.5]
 [15.6]
 [13.9]
 [16.6]
 [14.8]
 [18.4]
 [21. ]
 [12.7]
 [14.5]
 [13.2]
 [13.1]
 [13.5]
 [18.9]
 [20. ]
 [21. ]
 [24.7]
 [30.8]
 [34.9]
 [26.6]
 [25.3]
 [24.7]
 [21.2]
 [19.3]
 [20. ]
 [16.6]
 [14.4]
 [19.4]
 [19.7]
 [20.5]
 [25. ]
 [23.4]
 [18.9]
 [35.4]
 [24.7]
 [31.6]
 [23.3]
 [19.6]
 [18.7]
 [16. ]
 [22.2]
 [25. ]
 [33. ]
 [23.5]
 [19.4]
 [22. ]
 [17.4]
 [20

33.	Создайте модель линейной регрессии с помощью функции LinearRegression() из модуля sklearn.linear_model. Результат запишите в переменную model.

In [59]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()


34.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'neg_mean_absolute_error'. Разбиение на фолды (cv) задайте с помощью функции KFold() из модуля sklearn.model_selection, как в п. 18. Результат запишите в переменную results.

In [62]:
from sklearn.model_selection import KFold
kf=KFold(n_splits=10, shuffle=True, random_state=7)
results = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_absolute_error') 
results

array([-3.24826136, -4.11939173, -2.27347659, -4.17287637, -3.8034257 ,
       -3.20177954, -4.13777416, -3.11898137, -3.49219054, -2.30192009])

35.	Выведите на экран среднюю величину средней абсолютной ошибки (MAE), взятую с отрицательным знаком, и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std(). Прокомментируйте полученный результат.

In [63]:
print(results.mean(), results.std())

-3.387007745115897 0.6666977115119648


36.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'neg_mean_squared_error'. Разбиение на фолды (cv) задайте, как в п. 18. Результат запишите в переменную results.
37.	Выведите на экран среднюю величину средней квадратичной ошибки (MSE), взятую с отрицательным знаком, и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std(). Прокомментируйте полученный результат.


In [65]:
kf=KFold(n_splits=10, shuffle=True, random_state=7)
results = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error') 
print(results)
print(results.mean(), results.std())

[-16.63108562 -46.479212    -8.88386462 -30.04919184 -33.73045179
 -19.44900935 -32.03771938 -18.38474957 -22.04448736  -9.77524658]
-23.746501811313443 11.143430110698068


38.	Проведите кросс-валидацию, воспользовавшись функцией cross_val_score(model, X, Y, cv=kfold, scoring=scoring) из модуля sklearn.model_selection. В качестве оценочной функции (scoring) укажите 'r2'. Разбиение на фолды (cv) задайте, как в п. 18. Результат запишите в переменную results. 
39.	Выведите на экран среднюю величину коэффициента детерминации (R^2) и стандартное отклонение полученных оценок, воспользовавшись, соответственно, функциями mean() и std(). Прокомментируйте полученный результат.


In [66]:
results = cross_val_score(model, X, y, cv=kf, scoring='r2') 
print(results)
print(results.mean(), results.std())

[0.74116081 0.51162616 0.84734771 0.74188702 0.61894709 0.63591017
 0.69113678 0.79098704 0.81603103 0.78664942]
0.7181683241114094 0.09866585171842564
